// controllers/country.js
var db = require('controllers/database');
var mongoose = require('mongoose');
var countrys = mongoose.model('countrys');
var marks = mongoose.model('marks');
var pubs = mongoose.model('pubs');

exports.getAllCountry = function(callback){
	/*countrys.find({},function(err,data){
		if(data.length != null){
			callback({'res':true,'data':data});
		}else{
			callback({'res':false});
		}
	});*/
	countrys.find({},{_id:0,name:1},function(err,data){
		if(data.length != null){
			callback({'res':true,'data':data});
		}else{
			callback({'res':false});
		}
	}).sort({name:1});
}

// TODO: add new country
exports.setCountry = function(callback){
  countrys.insert({user:'sal'},function(err){
		if(err){
			callback({'res':false});
		}else {
			callback({'res':true});
		}
  });
}

exports.getAllCity = function(name,callback){
	/*countrys.find({name:name},function(err,data){
		if(data.length != null){
			callback({'res':true,'data':data[0].city});
		}else{
			callback({'res':false});
		}
	});*/
	countrys.aggregate({$unwind:'$city'},{$match:{name:name}},{$sort:{'city.name':1}},function(err,data){
		if(data.length != null){
			var dataCity = [];
      for (i = 0; i < data.length; i++) {
        dataCity.push({ 'name':data[i].city.name });
      }
			callback({'res':true,'data':dataCity});
		}else{
			callback({'res':false});
		}
	});
}

exports.getAllMark = function(callback){
	marks.find({},{_id:0,name:1},function(err,data){
		if (data.length != 0) {
			callback({'res':true,'data':data});
		} else {
			callback({'res':false});
		}
	});
}

exports.getAllPublicity = function(callback){
  pubs.find({},function(err,doc){
		if (doc.length != 0) {
      callback({'res':true,'data':doc});
    } else {
      callback({'res':false});
    }
  });
}

exports.getOnePublicity = function(name,callback){
  pubs.find({name:{$regex:new RegExp(name.toLowerCase(), "i")}},function(err,doc){
		if (doc.length != 0) {
      callback({'res':true,'data':doc});
    } else {
      callback({'res':false});
    }
  });
}

exports.getPublicityById = function(id,callback){
  pubs.find({_id:id},function(err,doc){
		if (doc.length != 0) {
      callback({'res':true,'data':doc});
    } else {
      callback({'res':false});
    }
  });
}

exports.addPublicity = function(name,category,period,price,date,callback){
  var pr = 0;
  if(period === '6 month'){
    pr = parseInt(price);
  }else if(period === '1 year'){
    pr = parseInt(price) / 2;
  }else if(period === '5 years'){
    pr = parseInt(price) / 10;
  }
  var type = '';
  if (pr < 3000) {
    type = 'bad';
  } else if (pr >= 3000 && pr < 6000 ) {
    type = 'medium';
  } else if (pr > 6000) {
    type = 'good';
  }
  var newTaxi = new pubs({name:name,category:category,period:period,price:price,type:type,date:date});
  pubs.find({name:name},function(err,data){
    if(data.length == 0){
      newTaxi.save(function(err){
          if(err){
            callback({'res':false,'response':"Error Publicity, you have not added a new publicity."});
          }else{
            callback({'res':true,'response':"Successfully Publicity, you have added a new publicity."});
          }
        });
    }else{
      callback({'res':false,'response':"Error Publicity, the name already registered."});
    }
  });
}

exports.editPublicity = function(id,name,category,period,price,date,callback){
  var pr = 0;
  if(period === '6 month'){
    pr = parseInt(price);
  }else if(period === '1 year'){
    pr = parseInt(price) / 2;
  }else if(period === '5 years'){
    pr = parseInt(price) / 10;
  }
  var type = '';
  if (pr < 3000) {
    type = 'bad';
  } else if (pr >= 3000 && pr < 6000 ) {
    type = 'medium';
  } else if (pr >= 6000) {
    type = 'good';
  }
  pubs.update({_id:id},{name:name,category:category,period:period,price:price,type:type,date:date},function(err){
    if (err) {
      callback({'res':false,'response':"Error Publicity, you have not edited a publicity."});
    }else{
      callback({'res':true,'response':"Successfully Publicity, you have edited a publicity."});
    }
  });
}

exports.deletePublicity = function(id,callback){
  pubs.remove({_id:id},function(err,data){
    if (err) {
      callback({'res':false,'response':"Error Publicity, you have not deleted a publicity."});
    } else {
      callback({'res':true,'response':"Successfully Publicity, you have deleted a publicity."});
    }
  });
}
