// controllers/reclamation.js
var db = require('controllers/database');
var mongoose = require('mongoose');
var reclamations = mongoose.model('reclamations');
var method	= require('../../route/methods');

exports.getAllReclamation = function(token,callback){
  reclamations.find({token:token},function(err,data){
    if(data.length != 0){
      var dataRec = [];
      for (i = 0; i < data.length; i++) {
        dataRec.push({ '_id':data[i]._id, 'subject':data[i].subject, 'date':data[i].date, 'status':data[i].status, 'me':data[i].me });
      }
      callback({'res':true,'data':dataRec});
    }else{
      callback({'res':false});
    }
  });
}

exports.getAllReclamationAdmin = function(callback){
  reclamations.find({},function(err,data){
    if (data.length != 0) {
      var dataRec = [];
      for (i = 0; i < data.length; i++) {
        dataRec.push({ '_id':data[i]._id, 'subject':data[i].subject, 'date':data[i].date, 'status':data[i].status, 'me':data[i].me });
      }
      callback({'res':true,'data':dataRec});
    } else {
      callback({'res':false});
    }
  });
}

exports.addReclamation = function(token,username,subject,message,callback){
  var datenew = new Date();
  var date = datenew.getDate() + "/"
    + (datenew.getMonth()+1)  + "/"
    + datenew.getFullYear() + " "
    + datenew.getHours() + ":"
    + datenew.getMinutes() + ":"
    + datenew.getSeconds();
  var newReclamation = new reclamations({token:token,username:username,subject:subject,date:date,
    status:true,me:true,messages:{sender:true,message:message,date:date}});
  newReclamation.save(function(err){
    if (err) {
      callback({'res':false, 'response':"Error Reclamation, you have not added a new reclamation."});
    } else {
      callback({'res':true, 'response':"Successfully Reclamation, you have added a new reclamation."});
    }
  });
}

exports.getAllMessage = function(idRec,callback){
  reclamations.find({_id:idRec},function(err,data){
    if(data.length != 0){
      callback({'res':true,'data':data[0].messages});
    }else{
      callback({'res':false});
    }
  });
}

exports.addMessage = function(idRec,message,callback){
  var datenew = new Date();
  var date = datenew.getDate() + "-"
    + (datenew.getMonth()+1)  + "-"
    + datenew.getFullYear() + " "
    + datenew.getHours() + ":"
    + datenew.getMinutes() + ":"
    + datenew.getSeconds();
  var keyVitual = method.keyVirtual();
  var key = method.key(keyVitual);
  reclamations.update({_id:idRec},{$set:{status:true,me:true},$push:{messages:{sender:true,message:message,date:date}}},function(err){
    if(err){
      callback({'res':false});
    }else{
      callback({'res':true,'key':keyVitual,'date':method.encode(date,key)});
    }
  });
}

exports.addMessageAdmin = function(idRec,message,callback){
  var datenew = new Date();
  var date = datenew.getDate() + "-"
    + (datenew.getMonth()+1)  + "-"
    + datenew.getFullYear() + " "
    + datenew.getHours() + ":"
    + datenew.getMinutes() + ":"
    + datenew.getSeconds();
  reclamations.update({_id:idRec},{$set:{status:true,me:false},$push:{messages:{sender:false,message:message,date:date}}},function(err){
    if(err){
      callback({'res':false});
    }else{
      callback({'res':true,'date':date});
    }
  });
}
